crema/qvalues.py

```diff
@@ -132,3 +136,213 @@ def _fdr2qvalue(scores, fdr):
+
+    Returns
+    -------
+    pi0 : float
+        Estimated pi_zero.
+    """
+    num_lambda = 100
+    max_lambda = 0.5
+    num_boot = 100
+
+    # LOGGER.debug("pval_list=%s", pval_list)
+
+    n_pval = pval_list.size
+    lambda_list = []
+    pi0s_list = []
+    for idx in range(0, num_lambda):
+        cur_lambda = ((idx + 1) / num_lambda) * max_lambda
+
+        # Find the index of the first element in pval_list
+        # that is > lambda.
+        start = np.searchsorted(pval_list, cur_lambda)
+        W1 = n_pval - start
+        pi0 = W1 / n_pval / (1.0 - cur_lambda)
+
+        if pi0 > 0.0:
+            lambda_list.append(cur_lambda)
+            pi0s_list.append(pi0)
+
+    # LOGGER.debug("%s", lambda_list)
+    # LOGGER.debug("%s", pi0s_list)
+
+    assert len(pi0s_list) != 0, (
+        "Error in the input data: "
+        "too good separation between target and decoy PSMs."
+    )
+
+    minPi0 = min(pi0s_list)
+
+    mse_list = np.zeros(len(pi0s_list))
+    max_size = 1000
+    # Examine which lambda level that is most stable under bootstrap
+    for i in range(0, num_boot):
+        # Create an array of bootstrapped p-values, and sort in ascending order.
+        num_draw = min(n_pval, max_size)
+        pBoot_list = np.random.choice(pval_list, size=num_draw, replace=True)
+        pBoot_list.sort()
+
+        for idx in range(0, len(lambda_list)):
+            start = np.searchsorted(pBoot_list, lambda_list[idx])
+            W1 = num_draw - start
+            pi0Boot = W1 / num_draw / (1.0 - lambda_list[idx])
+
+            # Estimated mean-squared error
+            mse_list[idx] += (pi0Boot - minPi0) * (pi0Boot - minPi0)
+
+    # Which index did the iterator go?
+    minIdx = np.argmin(mse_list)
+
+    # LOGGER.debug(f"Estimated pi0=%f at lambda=%f (MSE=%f)", pi0s_list[minIdx], lambda_list[minIdx], mse_list[minIdx])
+
+    pi0 = max(min(pi0s_list[minIdx], 1.0), 0.0)
+    return pi0
+
+
+@nb.njit
+def calculate_mixmax_qval(target_scores, decoy_scores, pi0):
+    """ """
+    # assert pi0 >= 0 and pi0 < 1
+
+    num_targets = target_scores.shape[0]
+    num_decoys = decoy_scores.shape[0]
+
+    h_w_le_z = np.zeros(num_decoys + 1)  # histogram for N_{w<=z}
+    h_z_le_z = np.zeros(num_decoys + 1)  # histogram for N_{z<=z}
+
+    for i in range(0, num_decoys):
+        h_w_le_z[i] = np.searchsorted(
+            target_scores, decoy_scores[i], side="right"
+        )
+        h_z_le_z[i] = np.searchsorted(
+            decoy_scores, decoy_scores[i], side="right"
+        )
+
+    h_w_le_z[num_decoys] = num_targets
+    h_z_le_z[num_decoys] = num_decoys
+
+    fdrmod = np.zeros(num_targets)
+    estPx_lt_zj = 0.0
+    E_f1_mod_run_tot = 0.0
+    j = num_decoys - 1
+    k = num_targets - 1
+    n_z_ge_w = 0
+    n_w_ge_w = 0
+    prev_fdr = -1
+
+    for i in range(num_targets - 1, -1, -1):
+        while j >= 0 and decoy_scores[j] >= target_scores[i]:
+            cnt_w = h_w_le_z[j + 1]
+            cnt_z = h_z_le_z[j + 1]
```

Comment:
Using an offset index into this array seems to introduce an off-by-one error compared to the published MixMax algorithm.

Because `h_z_le_z[j + 1]` is initialized above as `np.searchsorted(decoy_scores, decoy_scores[j + 1], side="right")` the value of `cnt_z` will be one plus the index of the last occurrence of `decoy_scores[j + 1]` in `decoy_scores`, which is the number of decoy scores as bad or worse than `decoy_scores[j + 1]`. In the paper this would be notated `N_{w \le z}(j + 1)` but we should be using `N_{w \le z}(j)`. The same issue applies to `cnt_w`.

I believe that the approach of offsetting the index by one would correctly implement the inequality when using `side="left"`, assuming that there are no tied scores. Using `side="right"` should implement the inequality without offset indices and also support tied scores.

I've verified that modifying this function to remove the index offset reproduces the exact _q_-value estimates of a from-scratch implementation based on the paper's Algorithm 1.